File: app/src/main/java/com/example/mike/mp3player/client/activities/SearchResultActivity.java
package com.example.mike.mp3player.client.activities;

import android.app.SearchManager;
import android.app.SearchableInfo;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.support.v4.media.MediaBrowserCompat.MediaItem;
import android.util.Log;

import androidx.annotation.VisibleForTesting;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.mike.mp3player.LogTagger;
import com.example.mike.mp3player.R;
import com.example.mike.mp3player.client.MyGenericItemTouchListener;
import com.example.mike.mp3player.client.callbacks.search.SearchResultListener;
import com.example.mike.mp3player.client.views.adapters.SearchResultAdapter;
import com.example.mike.mp3player.commons.MediaItemType;
import com.example.mike.mp3player.commons.MediaItemUtils;

import java.util.List;

import javax.inject.Inject;

import static com.example.mike.mp3player.commons.Constants.MEDIA_ITEM;

public abstract class SearchResultActivity extends MediaActivityCompat implements SearchResultListener, LogTagger,
    MyGenericItemTouchListener.ItemSelectedListener {
    private String currentQuery;
    private RecyclerView recyclerView;
    private SearchResultAdapter searchResultAdapter;
    private SearchView searchView;
    private Toolbar toolbar;

    @Override
    String getWorkerId() {
        return "SRCH_RSLT_ACTVTY";
    }

    @Override
    public void onConnected() {
        super.onConnected();
        initialiseView(R.layout.activity_search_results);
        this.getMediaBrowserAdapter().registerSearchResultListener(this);
        handleIntent(getIntent());
    }

    @Override
    boolean initialiseView(int layoutId) {
        setContentView(layoutId);
        this.toolbar = findViewById(R.id.toolbar);
        this.setSupportActionBar(toolbar);
        this.searchView = findViewById(R.id.search_view);
        this.recyclerView = findViewById(R.id.result_recycle_view);
        this.recyclerView.setAdapter(searchResultAdapter);
        MyGenericItemTouchListener itemTouchListener =
            new MyGenericItemTouchListener(getApplicationContext(), this);
        this.recyclerView.addOnItemTouchListener(itemTouchListener);
        itemTouchListener.setParentView(recyclerView);
        this.recyclerView.setLayoutManager(new LinearLayoutManager(getApplicationContext()));

        // Get the SearchView and set the searchable configuration
        SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
        ComponentName componentName = new ComponentName(getApplicationContext(), SearchResultActivityInjector.class);
        SearchableInfo searchableInfo = searchManager.getSearchableInfo(componentName);
        //    Assumes current activity is the searchable activity
        this.getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        searchView.setSearchableInfo(searchableInfo);
        searchView.setIconifiedByDefault(false); // Do not iconify the widget; expand it by default
        searchView.setSubmitButtonEnabled(true);
        searchView.setBackgroundColor(Color.WHITE);
        return true;
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        handleIntent(intent);
    }

    @Override
    public void itemSelected(MediaItem item) {
        if (null != item) {
            Class<?> intentClass = null;
            MediaItemType mediaItemType = MediaItemUtils.getMediaItemType(item);
            switch (mediaItemType) {
                case SONG:
                case SONGS:
                    mediaControllerAdapter.playFromMediaId(MediaItemUtils.getLibraryId(item), null);
                    break;
                case FOLDER:
                case FOLDERS:intentClass = FolderActivityInjector.class;
                    break;
                default:
                    break;
            }
            if (null != intentClass) {
                Intent intent = new Intent(getApplicationContext(), intentClass);
                intent.putExtra(MEDIA_ITEM, item);
                startActivity(intent);
            }
        }

    }

    private void handleIntent(Intent intent) {
        if (Intent.ACTION_SEARCH.equals(intent.getAction())) {
            this.currentQuery = intent.getStringExtra(SearchManager.QUERY);
            if (null != searchView) {
                this.searchView.setQuery(currentQuery, false);
            }
            mediaBrowserAdapter.search(currentQuery, null);
        }
    }

    @Override
    public void onSearchResult(List<MediaItem> searchResults) {
        searchResultAdapter.getItems().clear();
        searchResultAdapter.getItems().addAll(searchResults);
        searchResultAdapter.notifyDataSetChanged();
        Log.i(getLogTag(), "received search results");
    }

    @Override
    public String getLogTag() {
        return "SRCH_RSLT_ACTVTY";
    }

    @Inject
    public void setSearchResultAdapter(SearchResultAdapter searchResultAdapter) {
        this.searchResultAdapter = searchResultAdapter;
    }

    @VisibleForTesting
    public SearchView getSearchView() {
        return searchView;
    }

    @VisibleForTesting
    public void setSearchView(SearchView searchView) {
        this.searchView = searchView;    }
}

File: app/src/main/java/com/github/goldy1992/mp3player/client/activities/SearchResultActivity.kt
package com.github.goldy1992.mp3player.client.activities

import android.app.SearchManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.support.v4.media.MediaBrowserCompat
import android.util.Log
import androidx.recyclerview.widget.LinearLayoutManager
import com.github.goldy1992.mp3player.LogTagger
import com.github.goldy1992.mp3player.R
import com.github.goldy1992.mp3player.client.MyGenericItemTouchListener
import com.github.goldy1992.mp3player.client.MyGenericItemTouchListener.ItemSelectedListener
import com.github.goldy1992.mp3player.client.callbacks.search.SearchResultListener
import com.github.goldy1992.mp3player.client.views.adapters.SearchResultAdapter
import com.github.goldy1992.mp3player.commons.Constants
import com.github.goldy1992.mp3player.commons.MediaItemType
import com.github.goldy1992.mp3player.commons.MediaItemUtils
import kotlinx.android.synthetic.main.activity_search_results.*
import javax.inject.Inject

abstract class SearchResultActivity : MediaActivityCompat(), SearchResultListener, LogTagger, ItemSelectedListener {

    private var currentQuery: String? = null
    private var searchResultAdapter: SearchResultAdapter? = null

    override val workerId: String
        get() = "SRCH_RSLT_ACTVTY"

    override fun onConnected() {
        super.onConnected()
        initialiseView(R.layout.activity_search_results)
        mediaBrowserAdapter?.registerSearchResultListener(this)
        handleIntent(intent)
    }

    public override fun initialiseView(layoutId: Int): Boolean {
        setContentView(layoutId)
        setSupportActionBar(toolbar)

        recyclerView.setAdapter(searchResultAdapter)
        val itemTouchListener = MyGenericItemTouchListener(applicationContext, this)
        recyclerView.addOnItemTouchListener(itemTouchListener)
        itemTouchListener.parentView =recyclerView
        recyclerView.setLayoutManager(LinearLayoutManager(applicationContext))
        // Get the SearchView and set the searchable configuration
        val searchManager = getSystemService(Context.SEARCH_SERVICE) as SearchManager
        val componentName = ComponentName(applicationContext, SearchResultActivityInjector::class.java)
        val searchableInfo = searchManager.getSearchableInfo(componentName)
        //    Assumes current activity is the searchable activity
        this.supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        searchView.setSearchableInfo(searchableInfo)
        searchView.setIconifiedByDefault(false) // Do not iconify the widget; expand it by default
        searchView.isSubmitButtonEnabled = true
        searchView.setBackgroundColor(Color.WHITE)
        return true
    }

    public override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        handleIntent(intent)
    }

    override fun itemSelected(item: MediaBrowserCompat.MediaItem?) {
        if (null != item) {
            var intentClass: Class<*>? = null
            val mediaItemType = MediaItemUtils.getMediaItemType(item)
            when (mediaItemType) {
                MediaItemType.SONG, MediaItemType.SONGS -> mediaControllerAdapter!!.playFromMediaId(MediaItemUtils.getLibraryId(item), null)
                MediaItemType.FOLDER, MediaItemType.FOLDERS -> intentClass = FolderActivityInjector::class.java
                else -> {
                }
            }
            if (null != intentClass) {
                val intent = Intent(applicationContext, intentClass)
                intent.putExtra(Constants.MEDIA_ITEM, item)
                startActivity(intent)
            }
        }
    }

    private fun handleIntent(intent: Intent) {
        if (Intent.ACTION_SEARCH == intent.action) {
            currentQuery = intent.getStringExtra(SearchManager.QUERY)
            if (null != searchView) {
                searchView!!.setQuery(currentQuery, false)
            }
            mediaBrowserAdapter!!.search(currentQuery, null)
        }
    }

    override fun onSearchResult(searchResults: List<MediaBrowserCompat.MediaItem>?) {
        searchResultAdapter!!.items.clear()
        searchResultAdapter!!.items.addAll(searchResults!!)
        searchResultAdapter!!.notifyDataSetChanged()
        Log.i(logTag, "received search results")
    }

    override fun getLogTag(): String {
        return "SRCH_RSLT_ACTVTY"
    }

    @Inject
    fun setSearchResultAdapter(searchResultAdapter: SearchResultAdapter?) {
        this.searchResultAdapter = searchResultAdapter
    }

}